stdlib/public/Observation/Sources/Observation/Observed.swift

```diff
@@ -0,0 +1,226 @@
+    // it is bound to be called on the specified isolation of the construction
+    fileprivate static func trackEmission(isolation trackingIsolation: isolated (any Actor)?, state: SharedState<State>, work: @escaping @Sendable () throws(Failure) -> Element?) throws(Failure) -> Element? {
+      // this ferries in an intermediate form with Result to skip over `withObservationTracking` not handling errors being thrown
+      // particularly this case is that the error is also an iteration state transition data point (it terminates the sequence)
+      // so we need to hold that to get a chance to catch and clean-up
+      let result = withObservationTracking {
+        Result(catching: work)
+      } onChange: { [state] in
+        // resume all cases where the awaiting continuations are awaiting a willSet
+        State.emitWillChange(state)
+      }
+      return try result.get()
+    }
+    
+    fileprivate mutating func terminate(throwing failure: Failure? = nil) throws(Failure) -> Element? {
+      // flag the sequence as terminal by nil'ing out the state
+      state = nil
+      if let failure {
+        throw failure
+      } else {
+        return nil
+      }
+    }
+    
+    public mutating func next(isolation iterationIsolation: isolated (any Actor)? = #isolation) async throws(Failure) -> Element? {
+      // early exit if the sequence is terminal already
+      guard let state else { return nil }
+      do {
+        // there are two versions;
+        // either the tracking has never yet started at all and we need to prime the pump
+        // or the tracking has already started and we are going to await a change
+        if State.startTracking(state) {
+          guard !Task.isCancelled else { return nil }
+          // start by directly tracking the emission via a withObservation tracking on the isolation specified fro mthe init
+          guard let element = try await Iterator.trackEmission(isolation: isolation, state: state, work: work) else {
+            // the user returned a nil from the closure so terminate the sequence
+            return try terminate()
+          }
+          return element
+        } else {
+          // set up an id for this generation
+          let id = State.generation(state)
+          // wait for the willChange (and NOT the value itself)
+          // since this is going to be on the isolation of the object (e.g. the isolation specified in the initialization)
+          // this will mean our next await for the emission will ensure the suspension return of the willChange context
+          // back to the trailing edges of the mutations. In short, this enables the transactionality bounded by the
+          // isolation of the mutation.
+          await withTaskCancellationHandler {
+            await State.willChange(state, id: id)
+          } onCancel: {
+            // ensure to clean out our continuation uon cancellation
+            State.cancel(state, id: id)
+          }
```

Comment:
i'm having trouble convincing myself this can't potentially leak a continuation. e.g. if the ordering is:

1. `willChange` begins executing and runs up to just before its critical region
2. the Task is cancelled concurrently & `State.cancel` fully executes, removing the not-yet-registered continuation
3. `willChange` enters its critical region and registers a continuation for the id that was just cancelled

i guess if `emitWillChange()` eventually runs again, then at that point everything will be flushed, but maybe if this situation is possible we'd be stuck awaiting the `willChange` until then?

do we need to perform some additional state checks within the critical regions to avoid this, or can the scenario be ruled out analytically?

Reply:
There was a tiny window that could happen; I pushed a change that should easily handle that case.